File: app/src/main/java/com/xrevival/gltools26/core/GLUtils.java
package com.gl.revival;

import android.opengl.GLES20;
import android.opengl.GLES30;
import javax.microedition.khronos.egl.EGL10;
import javax.microedition.khronos.egl.EGLContext;
import javax.microedition.khronos.egl.EGLDisplay;

public final class GLUtils {
    private GLUtils() {}

    public static String glGetStringNative(int name) {
        return GLES20.glGetString(name);
    }

    public static int parseGlVersionMajor(String version) {
        if (version == null || version.isEmpty()) return 2;
        try {
            String[] parts = version.trim().split("\s+");
            String num = parts[0];
            if (num.startsWith("OpenGL ES ")) num = num.substring(10);
            return Integer.parseInt(num.substring(0, 1));
        } catch (Exception e) {
            GLLog.w("GLUtils", "parseGlVersionMajor fallback: " + e.getMessage());
            return 2;
        }
    }

    public static boolean isEglContextReady() {
        try {
            EGL10 egl = (EGL10) javax.microedition.khronos.egl.EGLContext.getEGL();
            EGLDisplay dpy = egl.eglGetDisplay(EGL10.EGL_DEFAULT_DISPLAY);
            return dpy != null && dpy != EGL10.EGL_NO_DISPLAY;
        } catch (Exception e) {
            return false;
        }
    }

    public static String sanitizeRenderer(String raw) {
        if (raw == null) return "Unknown";
        return raw.replaceAll("[^a-zA-Z0-9 \-_()/.]", "").trim();
    }

    public static int clamp(int v, int min, int max) {
        return Math.max(min, Math.min(max, v));
    }

    public static float clampf(float v, float min, float max) {
        return Math.max(min, Math.min(max, v));
    }

    public static int nextPowerOfTwo(int x) {
        if (x <= 1) return 1;
        return Integer.highestOneBit(x - 1) << 1;
    }

    public static String hex(int color) {
        return String.format("#%08X", color);
    }
}

File: src/main/java/com/gl/revival/GLUtils.kt
package com.gl.revival

import android.opengl.GLES20
import javax.microedition.khronos.egl.EGL10
import javax.microedition.khronos.egl.EGLContext
import javax.microedition.khronos.egl.EGLDisplay

object GLUtils {
    @JvmStatic
    fun glGetStringNative(name: Int): String = GLES20.glGetString(name)

    @JvmStatic
    fun parseGlVersionMajor(version: String?): Int {
        if (version.isNullOrEmpty()) return 2
        return try {
            val parts = version.trim().split("\s+".toRegex())
            val num = if (parts[0].startsWith("OpenGL ES ")) parts[0].substring(10) else parts[0]
            num.substring(0, 1).toInt()
        } catch (e: Exception) {
            GLLog.w("GLUtils", "parseGlVersionMajor fallback: ${e.message}")
            2
        }
    }

    @JvmStatic
    fun isEglContextReady(): Boolean =
        try {
            val egl = EGLContext.getEGL() as EGL10
            val dpy = egl.eglGetDisplay(EGL10.EGL_DEFAULT_DISPLAY)
            dpy != null && dpy != EGL10.EGL_NO_DISPLAY
        } catch (_: Exception) { false }

    @JvmStatic
    fun sanitizeRenderer(raw: String?): String =
        raw?.replace("[^a-zA-Z0-9 \-_()/.]".toRegex(), "")?.trim() ?: "Unknown"

    @JvmStatic
    fun clamp(v: Int, min: Int, max: Int): Int = maxOf(min, minOf(max, v))

    @JvmStatic
    fun clampf(v: Float, min: Float, max: Float): Float = maxOf(min, minOf(max, v))

    @JvmStatic
    fun nextPowerOfTwo(x: Int): Int =
        if (x <= 1) 1 else Integer.highestOneBit(x - 1) shl 1

    @JvmStatic
    fun hex(color: Int): String = String.format("#%08X", color)
}
